controllers/hvpa_controller.go

```diff
@@ -330,120 +344,85 @@ func (r *HvpaReconciler) reconcileVpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.
-
-	foundVpa := &vpa_api.VerticalPodAutoscaler{}
-	err = r.Get(context.TODO(), types.NamespacedName{Name: vpa.Name, Namespace: vpa.Namespace}, foundVpa)
-	if err != nil && errors.IsNotFound(err) {
-		if *updatePolicy == autoscalingv1alpha1.UpdateModePurge {
-			// If update policy is "Purge", then return
-			return nil, nil
-		}
-		log.V(2).Info("Creating VPA", "namespace", vpa.Namespace, "name", vpa.Name)
-		err = r.Create(context.TODO(), vpa)
+	if err := r.Create(context.TODO(), vpa); err != nil {
 		return nil, err
-	} else if err != nil {
-		return nil, err
-	} else if *updatePolicy == autoscalingv1alpha1.UpdateModePurge {
-		// Delete existing VPA, as VPA is not deployed in "Purge" mode
-		log.V(2).Info("Deleting VPA because update policy is set to Purge", "namespace", foundVpa.Namespace, "name", foundVpa.Name)
-		return nil, r.Delete(context.TODO(), foundVpa)
-	}
-
-	// Update the found object and write the result back if there are any changes
-	if !reflect.DeepEqual(vpa.Spec, foundVpa.Spec) {
-		foundVpa.Spec = vpa.Spec
-		log.V(2).Info("Updating VPA", "namespace", vpa.Namespace, "name", vpa.Name)
-		err = r.Update(context.TODO(), foundVpa)
-		if err != nil {
-			return nil, err
-		}
 	}
 
-	status := foundVpa.Status.DeepCopy()
-
-	return status, nil
+	return vpa.Status.DeepCopy(), nil
 }
 
 func (r *HvpaReconciler) reconcileHpa(hvpa *autoscalingv1alpha1.Hvpa) (*autoscaling.HorizontalPodAutoscalerStatus, error) {
-	metadata := hvpa.Spec.Hpa.Template.ObjectMeta.DeepCopy()
-
-	labels := metadata.GetLabels()
-	if labels == nil || len(labels) == 0 {
-		// TODO: Could be done better as part of validation
-		return nil, fmt.Errorf("Need labels in HPA template")
+	selector, err := metav1.LabelSelectorAsSelector(hvpa.Spec.Hpa.Selector)
+	if err != nil {
+		log.Error(err, "Error converting hpa selector to selector")
+		return nil, err
 	}
 
-	if ownerRef := metadata.GetOwnerReferences(); len(ownerRef) != 0 {
-		// TODO: Could be done better as part of validation
-		return nil, fmt.Errorf("hpa template in hvpa object already has an owner reference")
+	// list all hpas to include the hpas that don't match the hvpa`s selector
+	// anymore but has the stale controller ref.
+	hpas := &autoscaling.HorizontalPodAutoscalerList{}
+	err = r.List(context.TODO(), hpas, client.InNamespace(hvpa.Namespace))
+	if err != nil {
+		log.Error(err, "Error listing hpas")
+		return nil, err
 	}
 
-	if generateName := metadata.GetGenerateName(); len(generateName) != 0 {
-		log.V(3).Info("Warning", "Generate name provided in the hpa template will be ignored", generateName)
+	updatePolicy := hvpa.Spec.Hpa.UpdatePolicy.UpdateMode
+
+	// NOTE: filteredHpas are pointing to deepcopies of the cache, but this could change in the future.
+	// Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.2/pkg/cache/internal/cache_reader.go#L74
+	// if you need to modify them, you need to copy it first.
+	filteredHpas, err := r.claimHpas(hvpa, selector, hpas)
```

Comment:
Can we avoid this filtering by filtering in the `List` call [above](https://github.com/gardener/hvpa-controller/pull/34/files#diff-6354ac486a904d10093c497f09f83a3eR364)?